File: socnetscraper/parse.py
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any
from urllib.parse import quote

THREADS_HOSTS = ("https://www.threads.com", "https://www.threads.net")
POST_URL_RE = re.compile(
    r"https://www\.threads\.(?:com|net)/@([A-Za-z0-9._]+)/post/([A-Za-z0-9_-]+)"
)


def utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def from_unix(value: Any) -> str | None:
    try:
        ts = int(value)
    except (TypeError, ValueError):
        return None
    if ts > 10_000_000_000:
        ts //= 1000
    return datetime.fromtimestamp(ts, tz=timezone.utc).replace(microsecond=0).isoformat().replace(
        "+00:00", "Z"
    )


def search_url(query: str, filter_name: str) -> str:
    encoded = quote(query, safe="")
    url = f"https://www.threads.com/search?q={encoded}&serp_type=default"
    if filter_name == "recent":
        url += "&filter=recent"
    return url


# People often write anesthesia (наркоз) as "нархоз". Drop those unless the
# same text also talks about the university.
_ANESTHESIA = (
    "наркоз",
    "анестез",
    "эпидурал",
    "роддом",
    "босан",
    "кесар",
    "нархоз сал",
    "общий нархоз",
    "общ нархоз",
    "жартылай нархоз",
    "кс бол",
    "кс планов",
)
_UNIVERSITY = (
    "университет",
    "university",
    "студент",
    "грант",
    "магистр",
    "бакалавр",
    "бизнес шко",
    "narxoz",
    "mba",
    "оқи",
    "поступил",
    "препод",
    "семестр",
    "диплом",
    "кафедр",
    "факультет",
    "нархоз университет",
)


def mentions_narxoz(text: str | None, queries: list[str]) -> list[str]:
    blob = (text or "").casefold()
    hits = []
    for query in queries:
        needle = query.lstrip("#").casefold()
        if needle and needle in blob:
            hits.append(query)
    return hits


def is_university_mention(text: str | None, username: str | None, queries: list[str]) -> list[str]:
    hits = mentions_narxoz(text, queries)
    if not hits:
        return []
    user = (username or "").casefold()
    if any(name in user for name in ("narxoz", "нархоз")):
        return hits
    blob = (text or "").casefold()
    if any(word in blob for word in _UNIVERSITY):
        return hits
    if any(word in blob for word in _ANESTHESIA):
        return []
    return hits


def parse_post(item: dict[str, Any], query: str, source: str) -> dict[str, Any] | None:
    if not isinstance(item, dict):
        return None

    if item.get("permalink") or item.get("text") and item.get("username") and item.get("id"):
        url = item.get("permalink") or item.get("url")
        return _normalize(
            {
                "id": item.get("id"),
                "code": _code_from_url(url),
                "url": url,
                "username": item.get("username"),
                "text": item.get("text") or "",
                "published_at": item.get("timestamp") or item.get("published_at"),
                "like_count": item.get("like_count") or item.get("likeCount"),
                "reply_count": item.get("reply_count") or item.get("replyCount"),
                "repost_count": item.get("repost_count") or item.get("repostCount"),
                "query": query,
                "source": source,
            }
        )

    post = item.get("post") if isinstance(item.get("post"), dict) else item
    caption = post.get("caption") if isinstance(post.get("caption"), dict) else {}
    user = post.get("user") if isinstance(post.get("user"), dict) else {}
    code = post.get("code")
    username = user.get("username") or post.get("username")
    text = caption.get("text") or post.get("text") or ""
    published_at = from_unix(post.get("taken_at")) or post.get("published_at")
    threads_info = post.get("text_post_app_info")
    is_threads = isinstance(threads_info, dict) or bool(post.get("taken_at"))
    # Instagram grid/reel objects show up in the same GraphQL tree but are not Threads posts.
    if not is_threads or not published_at:
        return None
    if not code and not text:
        return None
    url = None
    if username and code:
        url = f"https://www.threads.com/@{username}/post/{code}"
    reply = post.get("text_post_app_info") if isinstance(post.get("text_post_app_info"), dict) else {}
    reply_to = None
    if isinstance(reply.get("reply_to_author"), dict):
        reply_to = reply["reply_to_author"].get("username")
    quoted = reply.get("quoted_post") if isinstance(reply.get("quoted_post"), dict) else None
    parent = None
    if quoted:
        parent = parent_card(parse_post({"post": quoted}, query, source) or {})
        if not parent.get("text") and not parent.get("username"):
            parent = None
    return _normalize(
        {
            "id": post.get("id") or post.get("pk") or code,
            "code": code,
            "url": url,
            "username": username,
            "full_name": user.get("full_name") or "",
            "verified": bool(user.get("is_verified")),
            "avatar_url": _avatar_url(user),
            "image_urls": _image_urls(post),
            "text": text,
            "published_at": published_at,
            "like_count": post.get("like_count"),
            "reply_count": _as_int(reply.get("direct_reply_count") or item.get("view_replies_cta_string")),
            "repost_count": post.get("repost_count") or reply.get("repost_count"),
            "query": query,
            "source": source,
            "is_reply": bool(reply_to),
            "reply_to": reply_to,
            "parent": parent,
        }
    )


def _normalize(post: dict[str, Any]) -> dict[str, Any] | None:
    if not post.get("id") and not post.get("code") and not post.get("url"):
        return None
    post["text"] = (post.get("text") or "").strip()
    post["username"] = post.get("username") or ""
    post["scraped_at"] = utc_now()
    post["image_urls"] = [url for url in (post.get("image_urls") or []) if url]
    post["images"] = post.get("images") or []
    post["avatar"] = post.get("avatar") or ""
    post["is_reply"] = bool(post.get("is_reply") or post.get("reply_to") or post.get("parent"))
    if post.get("parent") and not isinstance(post["parent"], dict):
        post["parent"] = None
    for field in ("like_count", "reply_count", "repost_count"):
        post[field] = _as_int(post.get(field))
    if post.get("url"):
        post["url"] = post["url"].replace("threads.net", "threads.com")
    return post


def _avatar_url(user: dict[str, Any]) -> str | None:
    if not isinstance(user, dict):
        return None
    if user.get("profile_pic_url"):
        return str(user["profile_pic_url"])
    versions = user.get("hd_profile_pic_versions")
    if isinstance(versions, list) and versions:
        last = versions[-1]
        if isinstance(last, dict) and last.get("url"):
            return str(last["url"])
    return None


def _best_image(node: dict[str, Any]) -> str | None:
    versions = node.get("image_versions2") if isinstance(node.get("image_versions2"), dict) else {}
    candidates = versions.get("candidates") or []
    if not candidates:
        return None
    best = max(candidates, key=lambda item: int(item.get("width") or 0))
    url = best.get("url")
    return str(url) if url else None


def _image_urls(post: dict[str, Any]) -> list[str]:
    urls: list[str] = []
    carousel = post.get("carousel_media")
    if isinstance(carousel, list) and carousel:
        for media in carousel:
            if isinstance(media, dict):
                url = _best_image(media)
                if url:
                    urls.append(url)
    else:
        url = _best_image(post)
        if url:
            urls.append(url)
    return list(dict.fromkeys(urls))


def _code_from_url(url: str | None) -> str | None:
    if not url:
        return None
    match = POST_URL_RE.search(url)
    return match.group(2) if match else None


def _as_int(value: Any) -> int | None:
    if value is None or value == "":
        return None
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        digits = re.split(r"\s+", value.strip())[0].replace(",", "")
        if digits.isdigit():
            return int(digits)
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def walk_posts(payload: Any) -> list[dict[str, Any]]:
    found: list[dict[str, Any]] = []
    seen: set[int] = set()

    def visit(node: Any) -> None:
        if id(node) in seen:
            return
        if isinstance(node, dict):
            seen.add(id(node))
            raw = node.get("post") if isinstance(node.get("post"), dict) else None
            if raw and raw.get("code") and (raw.get("taken_at") or raw.get("text_post_app_info")):
                found.append(node)
            elif (
                node.get("code")
                and node.get("taken_at")
                and isinstance(node.get("user"), dict)
            ):
                found.append({"post": node})
            for value in node.values():
                visit(value)
        elif isinstance(node, list):
            seen.add(id(node))
            for value in node:
                visit(value)

    visit(payload)
    return found


def urls_from_html(html: str) -> list[tuple[str, str]]:
    return list(dict.fromkeys(POST_URL_RE.findall(html)))


def parent_card(post: dict[str, Any]) -> dict[str, Any]:
    return {
        "code": post.get("code") or "",
        "username": post.get("username") or "",
        "text": post.get("text") or "",
        "url": post.get("url") or "",
        "avatar": post.get("avatar") or "",
        "avatar_url": post.get("avatar_url") or "",
        "images": post.get("images") or [],
        "image_urls": post.get("image_urls") or [],
        "published_at": post.get("published_at"),
        "verified": bool(post.get("verified")),
        "like_count": post.get("like_count") or 0,
        "reply_count": post.get("reply_count") or 0,
        "repost_count": post.get("repost_count") or 0,
    }


def walk_thread_groups(payload: Any) -> list[list[dict[str, Any]]]:
    groups: list[list[dict[str, Any]]] = []
    seen: set[int] = set()

    def visit(node: Any) -> None:
        if id(node) in seen:
            return
        if isinstance(node, dict):
            seen.add(id(node))
            items = node.get("thread_items")
            if isinstance(items, list) and items:
                groups.append(items)
            for value in node.values():
                visit(value)
        elif isinstance(node, list):
            seen.add(id(node))
            for value in node:
                visit(value)

    visit(payload)
    return groups


def parse_thread_group(items: list[Any], query: str, source: str) -> list[dict[str, Any]]:
    parsed: list[dict[str, Any]] = []
    for item in items:
        post = parse_post(item, query, source)
        if post:
            parsed.append(post)
    if len(parsed) >= 2:
        card = parent_card(parsed[0])
        for child in parsed[1:]:
            child["is_reply"] = True
            child["reply_to"] = child.get("reply_to") or parsed[0].get("username")
            if not (child.get("parent") or {}).get("text"):
                child["parent"] = card
    return parsed


def ingest_payload(payload: Any, query: str, source: str, posts: dict[str, dict[str, Any]]) -> None:
    grouped_codes: set[str] = set()
    for group in walk_thread_groups(payload):
        for post in parse_thread_group(group, query, source):
            _merge_parsed(posts, post)
            if post.get("code"):
                grouped_codes.add(str(post["code"]))
    for raw in walk_posts(payload):
        post = parse_post(raw, query, source)
        if post and str(post.get("code") or "") not in grouped_codes:
            _merge_parsed(posts, post)


def _merge_parsed(posts: dict[str, dict[str, Any]], parsed: dict[str, Any]) -> None:
    key = str(parsed.get("id") or parsed.get("code") or parsed.get("url") or "")
    if not key:
        return
    current = posts.get(key)
    if current is None and parsed.get("code"):
        code = str(parsed["code"])
        for existing_key, existing in posts.items():
            if str(existing.get("code") or "") == code:
                current = existing
                key = existing_key
                break
    if current is None:
        posts[key] = parsed
        return
    if parsed.get("text") and not current.get("text"):
        current["text"] = parsed["text"]
    if parsed.get("image_urls") and not current.get("image_urls"):
        current["image_urls"] = parsed["image_urls"]
    if parsed.get("avatar_url") and not current.get("avatar_url"):
        current["avatar_url"] = parsed["avatar_url"]
        current["verified"] = parsed.get("verified")
    if parsed.get("parent") and not (current.get("parent") or {}).get("text"):
        current["parent"] = parsed["parent"]
        current["is_reply"] = True
    if parsed.get("reply_to") and not current.get("reply_to"):
        current["reply_to"] = parsed["reply_to"]
        current["is_reply"] = True
    posts[key] = current

File: plugins/mcp/src/scraper/parse.ts
export type Post = Record<string, unknown>;

const POST_URL_RE =
  /https:\/\/www\.threads\.(?:com|net)\/@([A-Za-z0-9._]+)\/post\/([A-Za-z0-9_-]+)/g;

const ANESTHESIA = [
  "наркоз",
  "анестез",
  "эпидурал",
  "роддом",
  "босан",
  "кесар",
  "нархоз сал",
  "общий нархоз",
  "общ нархоз",
  "жартылай нархоз",
  "кс бол",
  "кс планов",
];

const UNIVERSITY = [
  "университет",
  "university",
  "студент",
  "грант",
  "магистр",
  "бакалавр",
  "бизнес шко",
  "narxoz",
  "mba",
  "оқи",
  "поступил",
  "препод",
  "семестр",
  "диплом",
  "кафедр",
  "факультет",
  "нархоз университет",
];

export function utcNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function fromUnix(value: unknown): string | undefined {
  const n = Number(value);
  if (!Number.isFinite(n) || n <= 0) {
    return undefined;
  }
  const seconds = n > 10_000_000_000 ? n / 1000 : n;
  return new Date(seconds * 1000).toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function searchUrl(query: string, filterName: string): string {
  const url = `https://www.threads.com/search?q=${encodeURIComponent(query)}&serp_type=default`;
  return filterName === "recent" ? `${url}&filter=recent` : url;
}

export function mentionsNarxoz(text: string | undefined, queries: string[]): string[] {
  const blob = (text || "").toLowerCase();
  return queries.filter((query) => {
    const needle = query.replace(/^#/, "").toLowerCase();
    return Boolean(needle) && blob.includes(needle);
  });
}

export function isUniversityMention(
  text: string | undefined,
  username: string | undefined,
  queries: string[]
): string[] {
  const hits = mentionsNarxoz(text, queries);
  if (!hits.length) {
    return [];
  }
  const user = (username || "").toLowerCase();
  if (user.includes("narxoz") || user.includes("нархоз")) {
    return hits;
  }
  const blob = (text || "").toLowerCase();
  if (UNIVERSITY.some((word) => blob.includes(word))) {
    return hits;
  }
  if (ANESTHESIA.some((word) => blob.includes(word))) {
    return [];
  }
  return hits;
}

function asInt(value: unknown): number | undefined {
  if (value === null || value === undefined || value === "") {
    return undefined;
  }
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.trunc(value);
  }
  const digits = String(value).trim().split(/\s+/)[0]?.replace(/,/g, "");
  if (digits && /^\d+$/.test(digits)) {
    return Number(digits);
  }
  const n = Number(value);
  return Number.isFinite(n) ? Math.trunc(n) : undefined;
}

function codeFromUrl(url: string | undefined): string | undefined {
  if (!url) {
    return undefined;
  }
  const match = /\/post\/([A-Za-z0-9_-]+)/.exec(url);
  return match?.[1];
}

function avatarUrl(user: Record<string, unknown>): string | undefined {
  if (user.profile_pic_url) {
    return String(user.profile_pic_url);
  }
  const versions = user.hd_profile_pic_versions;
  if (Array.isArray(versions) && versions.length) {
    const last = versions[versions.length - 1] as Record<string, unknown>;
    if (last?.url) {
      return String(last.url);
    }
  }
  return undefined;
}

function bestImage(node: Record<string, unknown>): string | undefined {
  const versions = node.image_versions2 as Record<string, unknown> | undefined;
  const candidates = (versions?.candidates as Record<string, unknown>[] | undefined) || [];
  if (!candidates.length) {
    return undefined;
  }
  const best = candidates.reduce((a, b) =>
    Number(a.width || 0) >= Number(b.width || 0) ? a : b
  );
  return best.url ? String(best.url) : undefined;
}

function imageUrls(post: Record<string, unknown>): string[] {
  const urls: string[] = [];
  const carousel = post.carousel_media;
  if (Array.isArray(carousel) && carousel.length) {
    for (const media of carousel) {
      if (media && typeof media === "object") {
        const url = bestImage(media as Record<string, unknown>);
        if (url) {
          urls.push(url);
        }
      }
    }
  } else {
    const url = bestImage(post);
    if (url) {
      urls.push(url);
    }
  }
  return [...new Set(urls)];
}

export function parentCard(post: Post): Post {
  return {
    code: post.code || "",
    username: post.username || "",
    text: post.text || "",
    url: post.url || "",
    avatar: post.avatar || "",
    avatar_url: post.avatar_url || "",
    images: post.images || [],
    image_urls: post.image_urls || [],
    published_at: post.published_at,
    verified: Boolean(post.verified),
    like_count: post.like_count || 0,
    reply_count: post.reply_count || 0,
    repost_count: post.repost_count || 0,
  };
}

function normalize(post: Post): Post | undefined {
  if (!post.id && !post.code && !post.url) {
    return undefined;
  }
  post.text = String(post.text || "").trim();
  post.username = post.username || "";
  post.scraped_at = utcNow();
  post.image_urls = ((post.image_urls as string[]) || []).filter(Boolean);
  post.images = post.images || [];
  post.avatar = post.avatar || "";
  post.is_reply = Boolean(post.is_reply || post.reply_to || post.parent);
  if (post.parent && typeof post.parent !== "object") {
    post.parent = undefined;
  }
  for (const field of ["like_count", "reply_count", "repost_count"] as const) {
    post[field] = asInt(post[field]);
  }
  if (typeof post.url === "string") {
    post.url = post.url.replace("threads.net", "threads.com");
  }
  return post;
}

export function parsePost(item: unknown, query: string, source: string): Post | undefined {
  if (!item || typeof item !== "object") {
    return undefined;
  }
  const rec = item as Record<string, unknown>;
  if (rec.permalink || (rec.text && rec.username && rec.id)) {
    const url = String(rec.permalink || rec.url || "");
    return normalize({
      id: rec.id,
      code: codeFromUrl(url),
      url,
      username: rec.username,
      text: rec.text || "",
      published_at: rec.timestamp || rec.published_at,
      like_count: rec.like_count || rec.likeCount,
      reply_count: rec.reply_count || rec.replyCount,
      repost_count: rec.repost_count || rec.repostCount,
      query,
      source,
    });
  }

  const post = (rec.post && typeof rec.post === "object" ? rec.post : rec) as Record<string, unknown>;
  const caption = (post.caption && typeof post.caption === "object" ? post.caption : {}) as Record<
    string,
    unknown
  >;
  const user = (post.user && typeof post.user === "object" ? post.user : {}) as Record<string, unknown>;
  const code = post.code as string | undefined;
  const username = String(user.username || post.username || "");
  const text = String(caption.text || post.text || "");
  const publishedAt = fromUnix(post.taken_at) || (post.published_at as string | undefined);
  const threadsInfo = post.text_post_app_info;
  const isThreads = Boolean(threadsInfo && typeof threadsInfo === "object") || Boolean(post.taken_at);
  if (!isThreads || !publishedAt) {
    return undefined;
  }
  if (!code && !text) {
    return undefined;
  }
  const reply = (threadsInfo && typeof threadsInfo === "object" ? threadsInfo : {}) as Record<
    string,
    unknown
  >;
  const replyToAuthor = reply.reply_to_author as Record<string, unknown> | undefined;
  const replyTo = replyToAuthor && typeof replyToAuthor === "object" ? String(replyToAuthor.username || "") : undefined;
  let parent: Post | undefined;
  if (reply.quoted_post && typeof reply.quoted_post === "object") {
    parent = parentCard(parsePost({ post: reply.quoted_post }, query, source) || {});
    if (!parent.text && !parent.username) {
      parent = undefined;
    }
  }
  return normalize({
    id: post.id || post.pk || code,
    code,
    url: username && code ? `https://www.threads.com/@${username}/post/${code}` : undefined,
    username,
    full_name: user.full_name || "",
    verified: Boolean(user.is_verified),
    avatar_url: avatarUrl(user),
    image_urls: imageUrls(post),
    text,
    published_at: publishedAt,
    like_count: post.like_count,
    reply_count: asInt(reply.direct_reply_count || rec.view_replies_cta_string),
    repost_count: post.repost_count || reply.repost_count,
    query,
    source,
    is_reply: Boolean(replyTo),
    reply_to: replyTo || undefined,
    parent,
  });
}

function walk(node: unknown, visitObj: (obj: Record<string, unknown>) => void, seen: WeakSet<object>): void {
  if (!node || typeof node !== "object" || seen.has(node)) {
    return;
  }
  seen.add(node);
  if (Array.isArray(node)) {
    for (const value of node) {
      walk(value, visitObj, seen);
    }
    return;
  }
  const obj = node as Record<string, unknown>;
  visitObj(obj);
  for (const value of Object.values(obj)) {
    walk(value, visitObj, seen);
  }
}

export function walkPosts(payload: unknown): Post[] {
  const found: Post[] = [];
  walk(
    payload,
    (node) => {
      const raw = node.post && typeof node.post === "object" ? (node.post as Record<string, unknown>) : undefined;
      if (raw && raw.code && (raw.taken_at || raw.text_post_app_info)) {
        found.push(node);
      } else if (node.code && node.taken_at && node.user && typeof node.user === "object") {
        found.push({ post: node });
      }
    },
    new WeakSet()
  );
  return found;
}

export function walkThreadGroups(payload: unknown): unknown[][] {
  const groups: unknown[][] = [];
  walk(
    payload,
    (node) => {
      if (Array.isArray(node.thread_items) && node.thread_items.length) {
        groups.push(node.thread_items);
      }
    },
    new WeakSet()
  );
  return groups;
}

export function parseThreadGroup(items: unknown[], query: string, source: string): Post[] {
  const parsed: Post[] = [];
  for (const item of items) {
    const post = parsePost(item, query, source);
    if (post) {
      parsed.push(post);
    }
  }
  if (parsed.length >= 2) {
    const card = parentCard(parsed[0]);
    for (const child of parsed.slice(1)) {
      child.is_reply = true;
      child.reply_to = child.reply_to || parsed[0].username;
      const parent = child.parent as Post | undefined;
      if (!parent?.text) {
        child.parent = card;
      }
    }
  }
  return parsed;
}

function mergeParsed(posts: Record<string, Post>, parsed: Post): void {
  let key = String(parsed.id || parsed.code || parsed.url || "");
  if (!key) {
    return;
  }
  let current = posts[key];
  if (!current && parsed.code) {
    const code = String(parsed.code);
    for (const [existingKey, existing] of Object.entries(posts)) {
      if (String(existing.code || "") === code) {
        current = existing;
        key = existingKey;
        break;
      }
    }
  }
  if (!current) {
    posts[key] = parsed;
    return;
  }
  if (parsed.text && !current.text) {
    current.text = parsed.text;
  }
  if ((parsed.image_urls as string[])?.length && !(current.image_urls as string[])?.length) {
    current.image_urls = parsed.image_urls;
  }
  if (parsed.avatar_url && !current.avatar_url) {
    current.avatar_url = parsed.avatar_url;
    current.verified = parsed.verified;
  }
  const parsedParent = parsed.parent as Post | undefined;
  const currentParent = current.parent as Post | undefined;
  if (parsedParent && !currentParent?.text) {
    current.parent = parsedParent;
    current.is_reply = true;
  }
  if (parsed.reply_to && !current.reply_to) {
    current.reply_to = parsed.reply_to;
    current.is_reply = true;
  }
  posts[key] = current;
}

export function ingestPayload(
  payload: unknown,
  query: string,
  source: string,
  posts: Record<string, Post>
): void {
  const grouped = new Set<string>();
  for (const group of walkThreadGroups(payload)) {
    for (const post of parseThreadGroup(group, query, source)) {
      mergeParsed(posts, post);
      if (post.code) {
        grouped.add(String(post.code));
      }
    }
  }
  for (const raw of walkPosts(payload)) {
    const post = parsePost(raw, query, source);
    if (post && !grouped.has(String(post.code || ""))) {
      mergeParsed(posts, post);
    }
  }
}

export function payloadsFromHtml(html: string): unknown[] {
  const payloads: unknown[] = [];
  const marker = '<script type="application/json"';
  let start = 0;
  while (true) {
    const idx = html.indexOf(marker, start);
    if (idx < 0) {
      break;
    }
    const openEnd = html.indexOf(">", idx);
    const close = html.indexOf("</script>", openEnd);
    if (openEnd < 0 || close < 0) {
      break;
    }
    const raw = html.slice(openEnd + 1, close);
    start = close + 9;
    if (!raw.includes("thread_items") && !raw.includes('"caption"')) {
      continue;
    }
    try {
      payloads.push(JSON.parse(raw));
    } catch {
      continue;
    }
  }
  return payloads;
}

export { POST_URL_RE };
